File: iOS/SignInEverywhere/FacebookProvider.m
//
//  FacebookProvider.m
//  SignInEverywhere
//
// Copyright 2013 Google Inc. All Rights Reserved.
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

#import "AccountView.h"
#import "FacebookProvider.h"

@implementation FacebookProvider

- (id)init {
  self = [super init];
  if (self) {
    self.isUserAction = NO;
    // Initialise default permissions.
    self.permissions = [[NSArray alloc] initWithObjects:
                            @"email",
                            @"user_likes",
                            nil];
    // See if we can auto sign-in.
    [FBSession openActiveSessionWithReadPermissions:self.permissions
                                       allowLoginUI:NO
                                  completionHandler:^(FBSession *session,
                                                      FBSessionState state,
                                                      NSError *error) {
                                    [self sessionStateChanged:session
                                                        state:state
                                                        error:error];
                                  }];
  }
  return self;
}

- (NSString *)providerId {
  return @"Facebook";
}

- (UIView *)buttonWithFrame:(CGRect)frame {
  AccountView *av = [[AccountView alloc] initWithFrame:frame];

  if (FBSession.activeSession.isOpen) {
    FBProfilePictureView *uiv = [[FBProfilePictureView alloc]
        initWithProfileID:self.user.id
          pictureCropping:FBProfilePictureCroppingSquare];
    [av setAccount:self.user.name byProvider:[self providerId] withPicture:uiv];
  } else {
    UIImageView *uiv = [[UIImageView alloc] initWithImage:
        [UIImage imageNamed:@"f_logo.png"]];
    [av setAccount:nil byProvider:[self providerId] withPicture:uiv];
  }
  return av;
}

- (void)signOut {
  [FBSession.activeSession closeAndClearTokenInformation];
}

-(void)signIn {
  if (FBSession.activeSession.isOpen && self.user) {
    return [self returnUser];
  }
  self.isUserAction = YES;
  [FBSession openActiveSessionWithReadPermissions:self.permissions
                                     allowLoginUI:YES
                                completionHandler:^(FBSession *session,
                                                    FBSessionState state,
                                                    NSError *error) {
                                    [self sessionStateChanged:session
                                                        state:state
                                                        error:error];
                                }];
}

- (void)disconnect {
  [self signOut];
}

- (BOOL)hasIdentity {
  return (FBSession.activeSession.isOpen && self.user);
}

- (void)returnUser {
  self.isUserAction = NO;
  User *u = [[User alloc] init];
  [u addAuthentication:self.user
            byProvider:[self providerId]
        withIdentifier:self.user.id
               andName:self.user.name];
  [[Authenticator sharedAuth] setUser:u];
}

/*
 * Callback for session changes.
 */
- (void)sessionStateChanged:(FBSession *)session
                      state:(FBSessionState) state
                      error:(NSError *)error {
  switch (state) {
    case FBSessionStateOpen:
      if (!error) {
        [[FBRequest requestForMe] startWithCompletionHandler:
         ^(FBRequestConnection *connection,
           NSDictionary<FBGraphUser> *user,
           NSError *error) {
           if (!error) {
             self.user = user;
             if (self.isUserAction) {
               [self returnUser];
             }
           }
         }];
      }
      break;
    case FBSessionStateClosed:
    case FBSessionStateClosedLoginFailed:
      [FBSession.activeSession closeAndClearTokenInformation];
      break;
    default:
      break;
  }

  if (error) {
    NSLog(@"Error with Facebook Login: %@", error);
  }
}

- (BOOL)openURL:(NSURL *)url
  sourceApplication:(NSString *)sourceApplication
         annotation:(id)annotation {
  // attempt to extract a token from the url
  return [FBSession.activeSession handleOpenURL:url];
}


@end

File: Android/src/com/example/signintest/providers/FacebookProvider.java
/*
 * Copyright 2013 Google Inc. All Rights Reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *     http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.example.signintest.providers;

import android.content.Intent;

import com.example.signintest.Provider;
import com.example.signintest.ProviderUtil;
import com.example.signintest.SignInFragment;
import com.example.signintest.SignInUser;
import com.facebook.*;
import com.facebook.model.*;

public class FacebookProvider implements 
	Provider, Session.StatusCallback, Request.GraphUserCallback {
	public static final String ID = "facebook";
	
	private SignInFragment mManager;
	private GraphUser mGraphUser;
	private boolean mIsSigningIn;
	
	@Override
	public String getId() {
		return FacebookProvider.ID;
	}

	@Override
	public void setFragment(SignInFragment manager) {
		mManager = manager;
		mIsSigningIn = false;
	}

	@Override
	public void trySilentAuthentication() {
		if(mGraphUser == null) {
			// Test login status - set allowLoginUI to false
			Session.openActiveSession(mManager.getActivity(), false, this);
		}
	}

	@Override
	public void signIn() {
		mIsSigningIn = true;
		if (mGraphUser != null) {
			returnUser();
			return;
		}
		// start Facebook Login - allowLoginUI is true
		Session.openActiveSession(mManager.getActivity(), true, this);
	}

	@Override
	public void signOut(SignInUser user) {
		Session.getActiveSession().closeAndClearTokenInformation();
		mGraphUser = null;
	}

	@Override
	public void disconnect(SignInUser user) {
		signOut(user);
	}
	
	public boolean hasFeature(ProviderUtil.Feature feature) {
		// This is just an example! Obviously specific features will 
		// depend on the APIs implemented etc.
		switch(feature) {
		case GRAPH:
		case FEATUREA:
		case PROFILE:
			return true;
		default: 
			return false;
		}
	}
	
	@Override
	public void detachFragment() {
		return;
	}

	@Override
	public boolean handleOnActivityResult(int requestCode, int resultCode,
			Intent data) {
		if(Session.getActiveSession() == null) {
			return false;
		}
		return Session.getActiveSession().onActivityResult(mManager.getActivity(), requestCode, resultCode, data);
	}

	@Override
	public String getUserId(Object user) {
		return ((GraphUser)user).getId();
	}

	/**
	 * Callbacks from API calls.
	 */
	@Override
	public void call(Session session, SessionState state, Exception exception) {
		if (session.isOpened()) {
			Request.executeMeRequestAsync(session, this);
		}
	}
	
	/**
	 * Callback after Graph API response with user object
	 */
	@Override
	public void onCompleted(GraphUser graphUser, Response response) {
		if (graphUser != null) {
			mGraphUser = graphUser;
			returnUser();
		}
	}
	
	/**
	 * Return the user to the manager.
	 */
	private void returnUser() {
		if(mIsSigningIn) {
			SignInUser user = mManager.buildSignInUser();
			user.setName(mGraphUser.getName());
			user.setProviderData(this, mGraphUser);
			mManager.onSignedIn(user);
			mIsSigningIn = false;
		}
	}
}
